File: Pract9_Hotel_Service/hotel_server.py
import Pyro4

#python -m Pyro4.naming

@Pyro4.expose
class HotelBookingSystem:

    def __init__(self):
        self.bookings = {}

    # Book room
    def book_room(self, guest_name):

        if guest_name in self.bookings:
            return f"Room already booked for {guest_name}"

        self.bookings[guest_name] = "Booked"

        return f"Room booked successfully for {guest_name}"

    # Cancel booking
    def cancel_booking(self, guest_name):

        if guest_name not in self.bookings:
            return f"No booking found for {guest_name}"

        del self.bookings[guest_name]

        return f"Booking cancelled for {guest_name}"

    # View all bookings
    def view_bookings(self):
        return self.bookings


def main():

    daemon = Pyro4.Daemon()

    ns = Pyro4.locateNS()

    hotel = HotelBookingSystem()

    uri = daemon.register(hotel)

    ns.register("hotel.booking", uri)

    print("Hotel Booking Server is Running...")
    print("URI:", uri)

    daemon.requestLoop()


if __name__ == "__main__":
    main()

File: pract9_Hotel_Service_Java/pract9/HotelServer.java
import java.rmi.RemoteException;
import java.rmi.registry.LocateRegistry;
import java.rmi.registry.Registry;
import java.rmi.server.UnicastRemoteObject;
import java.util.HashMap;

public class HotelServer extends UnicastRemoteObject implements HotelServiceInterface {

    // Stores booking information
    private HashMap<String, String> bookings = new HashMap<>();

    // Constructor
    protected HotelServer() throws RemoteException {
        super();
    }

    // Book a room
    @Override
    public String bookRoom(String guestName) throws RemoteException {

        if (bookings.containsKey(guestName)) {
            return "Room already booked for: " + guestName;
        }

        bookings.put(guestName, "Room Booked");

        return "Room successfully booked for: " + guestName;
    }

    // Cancel booking
    @Override
    public String cancelBooking(String guestName) throws RemoteException {

        if (!bookings.containsKey(guestName)) {
            return "No booking found for: " + guestName;
        }

        bookings.remove(guestName);

        return "Booking cancelled for: " + guestName;
    }

    // View all bookings
    @Override
    public String viewBookings() throws RemoteException {

        if (bookings.isEmpty()) {
            return "No current bookings.";
        }

        return "Current Bookings: " + bookings.keySet().toString();
    }

    // Main method
    public static void main(String[] args) {

        try {

            // Create server object
            HotelServer server = new HotelServer();

            // Create RMI registry on port 6000
            Registry registry = LocateRegistry.createRegistry(6001);

            // Bind server object with name
            registry.rebind("HotelService", server);

            System.out.println("Hotel Server is running and ready...");

        } catch (Exception e) {

            System.out.println("Server Error: " + e.getMessage());
        }
    }
}